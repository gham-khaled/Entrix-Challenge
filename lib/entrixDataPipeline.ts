import {Construct} from 'constructs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks'
import * as sfn from 'aws-cdk-lib/aws-stepfunctions'
import * as s3 from 'aws-cdk-lib/aws-s3'
import * as sns from 'aws-cdk-lib/aws-sns'
import {Duration} from 'aws-cdk-lib';
import {
    EventbridgeToStepfunctions,
    EventbridgeToStepfunctionsProps
} from '@aws-solutions-constructs/aws-eventbridge-stepfunctions';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import * as events from 'aws-cdk-lib/aws-events';

export class EntrixDataPipelineStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        // S3 Bucket
        const logBucket = new s3.Bucket(this, 'MyLogBucket');

        // Lambda Functions
        const lambdaA = new lambda.Function(this, 'lambdaA', {
            runtime: lambda.Runtime.PYTHON_3_9,
            handler: 'lambda_a.lambda_handler',
            code: lambda.Code.fromAsset('lib/lambda/dataPipeline/'),
        });
        const lambdaB = new lambda.Function(this, 'lambdaB', {
            runtime: lambda.Runtime.PYTHON_3_9,
            handler: 'lambda_b.lambda_handler',
            code: lambda.Code.fromAsset('lib/lambda/dataPipeline/'),
            environment: {
                LOG_BUCKET: logBucket.bucketName
            },
        });

        // SNS Topic
        const errorTopic = new sns.Topic(this, 'ErrorTopic');

        // SF Tasks
        const LambdaAJob = new tasks.LambdaInvoke(this, 'LambdaA', {
            lambdaFunction: lambdaA,
        });

        const publishError = new tasks.SnsPublish(this, 'Publish Error', {
            topic: errorTopic,
            message: sfn.TaskInput.fromJsonPathAt('$')
        });

        const LambdaBJob = new tasks.LambdaInvoke(this, 'LambdaB', {
            lambdaFunction: lambdaB,
        }).addCatch(publishError, {errors: [sfn.Errors.TASKS_FAILED]});

        const mapLambdaBJob = new sfn.Map(this, 'LambdaBMap', {
            itemsPath: sfn.JsonPath.stringAt('$.orders'),
        }).itemProcessor(LambdaBJob)

        const definition = LambdaAJob.next(new sfn.Choice(this, 'Results Empty?', {inputPath: "$.Payload"})
            .when(sfn.Condition.booleanEquals('$.results', false), LambdaAJob)
            .when(sfn.Condition.booleanEquals('$.results', true), mapLambdaBJob)
            .afterwards())


        // SF Cron
        const constructProps: EventbridgeToStepfunctionsProps = {
            stateMachineProps: {
                definition: definition
            },
            eventRuleProps: {
                schedule: events.Schedule.rate(Duration.hours(1))
            }
        };

        new EventbridgeToStepfunctions(this, 'hourlyDataPipeline', constructProps);

        // Permissions
        logBucket.grantPut(lambdaB);
        // errorTopic.addSubscription(new subscriptions.EmailSubscription('gham.khaled@gmail.com'));


    }

}