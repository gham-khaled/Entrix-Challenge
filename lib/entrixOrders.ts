import {Construct} from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as cdk from 'aws-cdk-lib';

export class EntrixOrdersStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // Dynamo DB Table
        const recordsTable = new dynamodb.Table(this, 'recordsTable', {
            partitionKey: {name: 'record_id', type: dynamodb.AttributeType.STRING},
            billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
        });

        // Lambda Function
        const postLambda = new lambda.Function(this, 'postLambda', {
            runtime: lambda.Runtime.PYTHON_3_9,
            handler: 'post_lambda.lambda_handler',
            code: lambda.Code.fromAsset('lib/lambda/recordsService'),
            environment: {
                DYNAMODB_TABLE_NAME: recordsTable.tableName,
                LOGLEVEL: "INFO",
            },
        });
        recordsTable.grantWriteData(postLambda)

        // API Gateway
        const api = new apigateway.RestApi(this, 'recordsAPI');


        const requestValidator = new apigateway.RequestValidator(this, 'BodyRequestValidator', {
            restApi: api,
            validateRequestBody: true,
        });

        const recordsRequestModel = api.addModel('RecordsRequestModel', {
            contentType: 'application/json',
            modelName: 'RecordsRequestModel',
            schema: {
                type: apigateway.JsonSchemaType.ARRAY,
                items: {
                    type: apigateway.JsonSchemaType.OBJECT,
                    properties: {
                        record_id: {type: apigateway.JsonSchemaType.STRING},
                    },
                    required: ['record_id'],
                },
            },
        });
        api.root.addMethod('POST',  new apigateway.LambdaIntegration(postLambda), {
            requestValidator: requestValidator,
            requestModels: {
                'application/json': recordsRequestModel,
            },
        });


    }

}
