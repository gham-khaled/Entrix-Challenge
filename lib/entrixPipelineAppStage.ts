import * as cdk from 'aws-cdk-lib';
import {Construct} from "constructs";
import {EntrixOrdersStack} from "./entrixOrders";
import {EntrixDataPipelineStack} from "./entrixDataPipeline";
import {CfnOutput} from "aws-cdk-lib";

export class EntrixPipelineAppStage extends cdk.Stage {
    public readonly urlOutput: CfnOutput;

    constructor(scope: Construct, id: string, props?: cdk.StageProps) {
        super(scope, id, props);
        const entrixOrders = new EntrixOrdersStack(this, 'EntrixOrders');
        new EntrixDataPipelineStack(this, 'EntrixOrders');
        this.urlOutput = entrixOrders.urlOutput;

    }
}