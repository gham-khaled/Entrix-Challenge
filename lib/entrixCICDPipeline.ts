import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {CodePipeline, CodePipelineSource, ShellStep} from 'aws-cdk-lib/pipelines';
import {EntrixPipelineAppStage} from "./entrixPipelineAppStage";

export class EntrixCICDPipeline extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const pipeline = new CodePipeline(this, 'Pipeline', {
            pipelineName: 'MyPipeline',
            synth: new ShellStep('Synth', {
                input: CodePipelineSource.gitHub('gham-khaled/Entrix-Challenge', 'master'),
                installCommands: ['npm ci', 'npm run build'],
                commands: ['npm run test', 'npx cdk synth']
            })
        });

        const devApp = new EntrixPipelineAppStage(this, "Dev", {env: {account: '830962405258', region: 'eu-west-1'}})

        const devStage = pipeline.addStage(devApp);

        devStage.addPost(new ShellStep("API Testing", {
            envFromCfnOutputs: {ENDPOINT_URL: devApp.urlOutput},
            commands: ['curl -Ssf  --location $ENDPOINT_URL \\\n' +
            '--header \'Content-Type: application/json\' \\\n' +
            '--data \'[\n' +
            ' {\n' +
            ' "record_id": "unique_id_1",\n' +
            ' "parameter_1": "abc",\n' +
            ' "parameter_2": 4\n' +
            ' },{\n' +
            ' "record_id": "unique_id_2",\n' +
            ' "parameter_1": "def",\n' +
            ' "parameter_2": 2.1\n' +
            ' }\n' +
            ']\'']

        }));

    }
}