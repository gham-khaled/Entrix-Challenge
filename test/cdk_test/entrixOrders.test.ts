import {EntrixOrdersStack} from "../../lib/entrixOrders";
import {Capture, Template} from "aws-cdk-lib/assertions";
import * as cdk from 'aws-cdk-lib';

test('DynamoDB Table Created', () => {
    const stack = new cdk.Stack();
    const entrixOrdersStack = new EntrixOrdersStack(stack, 'TestStack');
    const template = Template.fromStack(entrixOrdersStack);
    template.resourceCountIs("AWS::DynamoDB::Table", 1);
});
test('Lambda Created', () => {
    const stack = new cdk.Stack();
    const entrixOrdersStack = new EntrixOrdersStack(stack, 'TestStack');
    const template = Template.fromStack(entrixOrdersStack);
    template.resourceCountIs("AWS::Lambda::Function", 1);
});
test('API Gateway Created', () => {
    const stack = new cdk.Stack();
    const entrixOrdersStack = new EntrixOrdersStack(stack, 'TestStack');
    const template = Template.fromStack(entrixOrdersStack);
    template.resourceCountIs("AWS::ApiGateway::RestApi", 1);
});
test('Lambda Has Environment Variables', () => {
    const stack = new cdk.Stack();
    const entrixOrdersStack = new EntrixOrdersStack(stack, 'TestStack');

    // THEN
    const template = Template.fromStack(entrixOrdersStack);
    const envCapture = new Capture();
    template.hasResourceProperties("AWS::Lambda::Function", {
        Environment: envCapture,
    });

    expect(envCapture.asObject()).toEqual(
        {
            Variables: {
                DYNAMODB_TABLE_NAME: {
                    Ref: "recordsTable65B95BD2",
                },
            },
        }
    );
});
test('DynamoDB Billing mode is set to pay per request', () => {
    const stack = new cdk.Stack();
    const entrixOrdersStack = new EntrixOrdersStack(stack, 'TestStack');
    const template = Template.fromStack(entrixOrdersStack);
    template.hasResourceProperties('AWS::DynamoDB::Table', {
        BillingMode: "PAY_PER_REQUEST"
    });
});