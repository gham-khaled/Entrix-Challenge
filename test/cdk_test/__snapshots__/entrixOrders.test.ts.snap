// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing 1`] = `
{
  "Outputs": {
    "Url": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "recordsAPI217FECBD",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "recordsAPIDeploymentStageprodCE308A15",
            },
            "/",
          ],
        ],
      },
    },
    "recordsAPIEndpoint7CE9A2F5": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "recordsAPI217FECBD",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "recordsAPIDeploymentStageprodCE308A15",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "BodyRequestValidatorE3D5450A": {
      "Properties": {
        "RestApiId": {
          "Ref": "recordsAPI217FECBD",
        },
        "ValidateRequestBody": true,
      },
      "Type": "AWS::ApiGateway::RequestValidator",
    },
    "postLambdaE03F5E9F": {
      "DependsOn": [
        "postLambdaServiceRoleDefaultPolicy7705C00A",
        "postLambdaServiceRole3B04939D",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "8786ce2dbdb4878ec7a0a1eade0a6d39dcc941589ceb0af632067a175d8a20e9.zip",
        },
        "Environment": {
          "Variables": {
            "DYNAMODB_TABLE_NAME": {
              "Ref": "recordsTable65B95BD2",
            },
          },
        },
        "Handler": "post_lambda.lambda_handler",
        "Role": {
          "Fn::GetAtt": [
            "postLambdaServiceRole3B04939D",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
    "postLambdaServiceRole3B04939D": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "postLambdaServiceRoleDefaultPolicy7705C00A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "recordsTable65B95BD2",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "postLambdaServiceRoleDefaultPolicy7705C00A",
        "Roles": [
          {
            "Ref": "postLambdaServiceRole3B04939D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "recordsAPI217FECBD": {
      "Properties": {
        "Name": "recordsAPI",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "recordsAPIAccount915CE1E4": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "recordsAPI217FECBD",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "recordsAPICloudWatchRoleD6620155",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "recordsAPICloudWatchRoleD6620155": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "recordsAPIDeployment330D9AFF832a8624fe33e45c926e3ad2d6f41e34": {
      "DependsOn": [
        "BodyRequestValidatorE3D5450A",
        "recordsAPIPOST43C8F23A",
        "recordsAPIRecordsRequestModel2D9FD179",
      ],
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "recordsAPI217FECBD",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "recordsAPIDeploymentStageprodCE308A15": {
      "DependsOn": [
        "recordsAPIAccount915CE1E4",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "recordsAPIDeployment330D9AFF832a8624fe33e45c926e3ad2d6f41e34",
        },
        "RestApiId": {
          "Ref": "recordsAPI217FECBD",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "recordsAPIPOST43C8F23A": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "postLambdaE03F5E9F",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "RequestModels": {
          "application/json": {
            "Ref": "recordsAPIRecordsRequestModel2D9FD179",
          },
        },
        "RequestValidatorId": {
          "Ref": "BodyRequestValidatorE3D5450A",
        },
        "ResourceId": {
          "Fn::GetAtt": [
            "recordsAPI217FECBD",
            "RootResourceId",
          ],
        },
        "RestApiId": {
          "Ref": "recordsAPI217FECBD",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "recordsAPIPOSTApiPermissionEntrixOrdersStackrecordsAPI8B9EB264POST9B5070E1": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "postLambdaE03F5E9F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "recordsAPI217FECBD",
              },
              "/",
              {
                "Ref": "recordsAPIDeploymentStageprodCE308A15",
              },
              "/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "recordsAPIPOSTApiPermissionTestEntrixOrdersStackrecordsAPI8B9EB264POST078984F8": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "postLambdaE03F5E9F",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "recordsAPI217FECBD",
              },
              "/test-invoke-stage/POST/",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "recordsAPIRecordsRequestModel2D9FD179": {
      "Properties": {
        "ContentType": "application/json",
        "Name": "RecordsRequestModel",
        "RestApiId": {
          "Ref": "recordsAPI217FECBD",
        },
        "Schema": {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "items": {
            "properties": {
              "record_id": {
                "type": "string",
              },
            },
            "required": [
              "record_id",
            ],
            "type": "object",
          },
          "type": "array",
        },
      },
      "Type": "AWS::ApiGateway::Model",
    },
    "recordsTable65B95BD2": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "record_id",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "record_id",
            "KeyType": "HASH",
          },
        ],
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
